sw/device/lib/testing/meson.build

```diff
@@ -25,26 +25,50 @@ touch "@OUTDIR@/$EMPTY_FILE_FOR_DEP"
 echo "Done!"
 '''.format(sw_lib_testing_gtest_build_dir, sw_lib_testing_gtest_src_dir, sw_lib_testing_empty_file_for_dep)
 
-build_gtest = custom_target(
+sw_lib_testing_build_gtest = custom_target(
   'googletest',
   output: sw_lib_testing_empty_file_for_dep,
   depend_files: sw_lib_testing_gtest_lock_file,
-  command: ['bash', '-c', build_gtest_cmd],
+  command: ['bash', '-e', '-c', build_gtest_cmd],
+  console: true,
 )
 
-gtest_inc = include_directories(
-  '../../../vendor/google_googletest/googletest/include',
-  '../../../vendor/google_googletest/googlemock/include',
-)
-
-gtest = declare_dependency(
-  sources: [build_gtest],
-  include_directories: gtest_inc,
+sw_lib_testing_gtest = declare_dependency(
+  sources: [sw_lib_testing_build_gtest],
   link_args: [
     '-L' + sw_lib_testing_gtest_build_dir / 'lib',
     '-lgmock',
     '-lgmock_main',
     '-lgtest',
   ],
   dependencies: dependency('threads'),
+  compile_args: [
+    # These are necessary in order to make gtest headers correctly find other
+    # gtest headers.
+    '-I' + meson.source_root() / 'sw/vendor/google_googletest/googletest/include',
+    '-I' + meson.source_root() / 'sw/vendor/google_googletest/googlemock/include',
+  ],
 )
+
+sw_lib_testing_mock_mmio = declare_dependency(
+  link_with: static_library(
+    'mock_mmio',
+    sources: ['mock_mmio.cc'],
+    dependencies: [sw_lib_testing_gtest],
+    native: true,
+    cpp_args: ['-DMOCK_MMIO'],
+  )
+)
+
+# Example test using mock_mmio.h, which also serves to
+# test that mock_mmio.h works correctly.
+test('mock_mmio_test', executable(
```

Comment:
I have tried to add a quick and dirty code for `dif_uart` testing, however I ran into some issues. It might as well be that I am not doing the right thing, it would be nice to have a simple example of usage in terms of DIFs. Anyway what I have done is added the following block:
```
test('mock_dif_uart_test', executable(
  'mock_dif_uart_test',
  sources: [
    meson.source_root() / 'sw/device/lib/testing/dif/mock_dif_uart_test.cc',
    meson.source_root() / 'sw/device/lib/dif/dif_uart.c',
    [hw_ip_uart_reg_h],
  ],
  dependencies: [
    sw_lib_testing_gtest,
    sw_lib_testing_mock_mmio,
  ],
  native: true,
  cpp_args: ['-DMOCK_MMIO'],
))
```
When I build the project, I get the following errors:
```

/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `uart_tx_full':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:16: undefined reference to `mmio_region_get_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `uart_tx_empty':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:21: undefined reference to `mmio_region_get_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `uart_rx_empty':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:26: undefined reference to `mmio_region_get_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_watermark_rx_set':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:199: undefined reference to `mmio_region_nonatomic_set_mask32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_watermark_tx_set':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:233: undefined reference to `mmio_region_nonatomic_set_mask32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_irq_state_get':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:311: undefined reference to `mmio_region_get_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_irq_state_clear':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:330: undefined reference to `mmio_region_nonatomic_set_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_irq_enable':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:376: undefined reference to `mmio_region_nonatomic_set_bit32'
/usr/bin/ld: /home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:379: undefined reference to `mmio_region_nonatomic_clear_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_irq_force':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:393: undefined reference to `mmio_region_nonatomic_set_bit32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_rx_bytes_available':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:405: undefined reference to `mmio_region_read_mask32'
/usr/bin/ld: sw/device/lib/testing/871b391@@mock_dif_uart_test@exe/dif_uart.c.o: in function `dif_uart_tx_bytes_available':
/home/svt/development/repos/opentitan/sw/device/lib/dif/dif_uart.c:418: undefined reference to `mmio_region_read_mask32'
```
Is the problem because the named function prototypes have `inline` in them? As you have removed the `inline` qualifier for the mock functions, but not for the functions that are implemented in terms of mocks?

Reply:
I still think you need to pull in the libbase mmio library, too. I'll make it so mock_mmio does that automatically.

Reply:
I see, just tried this PR with your latest push, works nicely, thanks.